File: Java/subsets.java
//  Subsets

// Time O(2^N), Space O(N)
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> ans = new ArrayList<List<Integer>>();
        List<Integer> subset = new ArrayList<Integer>();
        dfs(nums, 0, subset, ans);
        return ans;
    }
    
    private void dfs(int[] nums, int start, List<Integer> subset, List<List<Integer>> ans) {
        ans.add(new ArrayList(subset));
        for (int i=start; i<nums.length; i++) {
            subset.add(nums[i]);
            dfs(nums, i+1, subset, ans);
            subset.remove(subset.size()-1);
        }
    }
}

File: leetcode/subsets.cc
//  Subsets

class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> res;
        vector<int> subset;
        dfs(nums, 0, subset, res);
        return res;
    }
    
    void dfs(vector<int> &nums, int cur, vector<int> &subset, vector<vector<int>> &res) {
        res.push_back(subset);
        if (cur == nums.size()) return;
        for (int i=cur; i<nums.size(); i++) {
            subset.push_back(nums[i]);
            dfs(nums, i+1, subset, res);
            subset.pop_back();
        }
    }
};